File: microservice/microservice-app/src/main/java/com/greg/microservice/web/CustomerController.java
package com.greg.microservice.web;

import com.greg.microservice.data.CustomerEntity;
import com.greg.microservice.data.CustomerRepo;
import com.greg.microsservice.shared.model.ApiResult;
import com.greg.microsservice.shared.model.CustomerCreateRequest;
import com.greg.microsservice.shared.model.CustomerModel;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@RestController
@RequestMapping("/customer")
@RequiredArgsConstructor
public class CustomerController {
    private final CustomerRepo customerRepo;

    @PostMapping
    public ApiResult createCustomer(CustomerCreateRequest request){
        var customer = new CustomerEntity();
        customer.setName(request.name());
        customerRepo.save(customer);
        return new ApiResult(customer.getId(), true ,"Customer created");
    }

    @GetMapping
    public List<CustomerModel> getCustomers(){
        var customers = StreamSupport.stream(customerRepo.findAll().spliterator(), true)
                .map(x-> new CustomerModel(x.getId(),x.getName(),x.getCreated()))
                .collect(Collectors.toList());

        return customers;
    }
}

File: microservice/microservice-app-kotlin/src/main/kotlin/com/greg/microserviceappkotlin/web/CustomerController.kt
package com.greg.microserviceappkotlin.web

import com.greg.microserviceappkotlin.data.CustomerEntity
import com.greg.microserviceappkotlin.data.CustomerRepo
import com.greg.microsservice.shared.model.ApiResult
import com.greg.microsservice.shared.model.CustomerCreateRequest
import com.greg.microsservice.shared.model.CustomerModel
import org.springframework.web.bind.annotation.*
import java.time.LocalDateTime
import java.util.*

@RestController
@RequestMapping("/customer")
class CustomerController(private val customerRepo: CustomerRepo) {

    @PostMapping
    fun createCustomer(@RequestBody request: CustomerCreateRequest): ApiResult {
        val customer = CustomerEntity(LocalDateTime.now(), UUID.randomUUID(),request.name)
        customerRepo.save(customer)
        return ApiResult(customer.id, true, "Customer created")
    }

    @GetMapping
    fun getAll(): List<CustomerModel>? {
       return customerRepo.findAll().asSequence().toList().stream().map { CustomerModel(it!!.id, it.name, it.created) }.toList()
    }
}